community/kernel/src/main/java/org/neo4j/kernel/impl/api/OffHeapCollectionsFactoryImpl.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import org.neo4j.collection.primitive.Primitive;
+import org.neo4j.collection.primitive.PrimitiveCollection;
+import org.neo4j.collection.primitive.PrimitiveIntObjectMap;
+import org.neo4j.collection.primitive.PrimitiveLongObjectMap;
+import org.neo4j.collection.primitive.PrimitiveLongSet;
+import org.neo4j.kernel.impl.util.diffsets.PrimitiveLongDiffSets;
+import org.neo4j.memory.MemoryAllocationTracker;
+import org.neo4j.memory.MemoryTracker;
+
+import static java.util.Objects.requireNonNull;
+import static org.neo4j.collection.primitive.PrimitiveLongCollections.emptySet;
+
+class OffHeapCollectionsFactoryImpl implements CollectionsFactory
+{
+    private final Collection<PrimitiveCollection> allocated = new ArrayList<>();
+    private final MemoryAllocationTracker memoryTracker;
+
+    OffHeapCollectionsFactoryImpl( MemoryAllocationTracker memoryTracker )
+    {
+        this.memoryTracker = requireNonNull( memoryTracker );
+    }
+
+    @Override
+    public PrimitiveLongSet newLongSet()
+    {
+        return track( Primitive.offHeapLongSet( memoryTracker ) );
+    }
+
+    @Override
+    public <V> PrimitiveLongObjectMap<V> newLongObjectMap()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <V> PrimitiveIntObjectMap<V> newIntObjectMap()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public PrimitiveLongDiffSets newLongDiffSets()
+    {
+        return new PrimitiveLongDiffSets( emptySet(), emptySet(), this );
+    }
+
+    @Override
+    public MemoryTracker getMemoryTracker()
+    {
+        return memoryTracker;
+    }
+
+    @Override
+    public void release()
+    {
+        allocated.forEach( PrimitiveCollection::close );
+        allocated.clear();
+    }
+
+    private <T extends PrimitiveCollection> T track( T collection )
+    {
+        allocated.add( collection );
```

Comment:
not quite sure we need to track all the allocated collections here; It would be useful for tests to see what collections were allocated but do not think its useful for production